validator_client/src/key_cache.rs

```diff
@@ -0,0 +1,342 @@
+use account_utils::create_with_600_perms;
+use bls::{Keypair, PublicKey};
+use eth2_keystore::json_keystore::{
+    Aes128Ctr, ChecksumModule, Cipher, CipherModule, Crypto, EmptyMap, EmptyString, KdfModule,
+    Sha256Checksum,
+};
+use eth2_keystore::{
+    decrypt, default_kdf, encrypt, keypair_from_secret, Error as KeystoreError, Uuid, IV_SIZE,
+    SALT_SIZE,
+};
+use rand::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::OpenOptions;
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+/// The file name for the serialized `KeyCache` struct.
+pub const CACHE_FILENAME: &str = "validator_key_cache.json";
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum State {
+    NotDecrypted,
+    DecryptedAndSaved,
+    DecryptedWithUnsavedUpdates,
+}
+
+fn not_decrypted() -> State {
+    State::NotDecrypted
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct KeyCache {
+    crypto: Option<Crypto>,
+    uuids: Vec<Uuid>,
+    #[serde(skip)]
+    pairs: HashMap<Uuid, Keypair>, //maps public keystore uuids to their corresponding Keypair
```

Comment:
I would have thought it would be simpler to key this via `PublicKey` instead of `Uuid`. An EIP-2335 has a plain-text `pubkey` field you can access without the password. Perhaps I'm missing something.

Reply:
The main idea here is to also consider the case that we have multiple keystores with the same PublicKey (but with different passwords and uuids, I assume if two keystores have the same uuid they are identical and also have the same password).

The goal is to keep the previous behavior (and basically don't make the change breaking), in this context especially this comment (https://github.com/sigp/lighthouse/blob/64e21c32eec426b418a947cdbbe460d35f64465e/validator_client/src/initialized_validators.rs#L486):

> A validator is considered "already known" and skipped if the public key is already known. I.e., if there are two different definitions with the same public key then the second will be ignored.

Currently I assume that two keystores with the same uuid are identical i.e. have the same secret and the same password ...